2단계 - 예외 처리 및 리팩터링

OikimiO opened the pull request:
안녕하세요 아라님
이번 스텝은 좀 더 발전된 모습을 보여야 하는데.. ㅠㅜ
생각보다 너무 급하게 만들어서 많이 엉성한 부분들이 보입니다.. ㅠㅜ

아래는 이전에 피드백을 참고하여 정리한 내용입니다. 

### ViewController로 비지니스 Controller와 분리
> 화면 응답용 컨트롤러와 api 응답용 컨트롤러를 나눠보는 건 어떤가요?

<img width="277" alt="image" src="https://github.com/next-step/spring-roomescape-reservation/assets/53805469/261db26d-2dc5-4bc6-ae46-9d590a139dec">

해당 조언에 따라 ViewController와 비지니스Controller로 구분해보았습니다

### 쿼리 메서드 명명 규칙으로 쿼리를 알기 쉽게..!
> readReservation() 은 단건 조회와 헷갈릴 수 있어서 (이번 요구사항에는 없지만)
목록 조회라는 내용이 더 드러나는 네이밍을 지어줘도 좋을 것 같아요 ㅎㅎ

쿼리 메서드 명명 규칙을 참고하여 구성을 변경했습니다.
조회를 변경하면서 삭제/저장 메서드도 변경했습니다 ㅎ

AS_IS > TO_BE

DTO read~: find~
read~: findAll
DTO save ~: save
void delet ~: delete

### ControllerTest, ServiceTest의 역할
<img width="328" alt="image" src="https://github.com/next-step/spring-roomescape-reservation/assets/53805469/7a23de76-5813-4799-8558-30532d5d4e85">

> 테스트 코드도 운영코드에 맞춰 클래스 분리가 필요하겠네요 ~!

<b>ControllerTest의 역할</b>
1) 각 Controller 호출을 통한 프로세스 흐름 파악
2) Controller에서 저장 호출시 데이터 유효성 검증 파악

<b>ServiceTest의 역할</b>
1) 각 도메인별 정책 파악

### ControllerTest - 저장 요청에 대한 유효성 검증 진행
<img width="436" alt="image" src="https://github.com/next-step/spring-roomescape-reservation/assets/53805469/865d1107-54e3-4ae8-9403-d2eb89391d35">

<img width="436" alt="image" src="https://github.com/next-step/spring-roomescape-reservation/assets/53805469/3b5f60b6-7941-4ce7-8192-188967af6a38">


<img width="436" alt="image" src="https://github.com/next-step/spring-roomescape-reservation/assets/53805469/f9b81731-db0e-476e-935e-a13e1a0c3467">


### ServiceTest - 지나간 날짜,시간 및 중복된 예약 예외 처리
<img width="436" alt="image" src="https://github.com/next-step/spring-roomescape-reservation/assets/53805469/8e47d338-9dad-4ce4-ab5e-2fab33ba639c">

<b>예외 처리</b>
발생되는 예외를 CustomException으로 처리하고 이를 @ControllerAdvice에서 ResponseEntity로 응답하도록 
구현하였습니다.

### ServiceTest - 외래키 사용에 따른 예외 처리
> 외래키를 사용해주셨군요 !
시간에 대한 예약이 존재하는 상황에서 시간을 삭제하려고 하면 어떻게 될까요? 🤔
2단계 예외처리 진행하시면서 같이 해결해볼 수 있을 것 같아요 ㅎㅎ
<img width="442" alt="image" src="https://github.com/next-step/spring-roomescape-reservation/assets/53805469/bfb9f1b8-9b3e-469f-9915-b2c816b2d8c1">
<img width="442" alt="image" src="https://github.com/next-step/spring-roomescape-reservation/assets/53805469/5bf1664e-2219-496c-9599-e5045a4f8d19">

<b>정책 - 예약에서 사용중인 예약시간&테마는 삭제하면 안돼요..!</b>

예약(Reservation)은 예약시간(ReservationTime)과 테마(Theme)에 대한 외래키를 가집니다. 
이로 인해 예약목록 중 예약시간,테마를 사용하고 있다면 외래키에 해당하는 값은 삭제할 수 없습니다.

<b>예외 처리</b>

해당 부분은 ... 당시엔 생각이 안나... 예외 처리를 진행하지 못했습니다..
생각을 정리하며 생각난건데.. 예외 처리를 Exception이 발생하는 영역에 
try with resource 형태로 구현하면 어떨까 싶습니다.

### 아쉬운 점
<b>1. ServiceTest에서 Controller로 요청을 하여 테스트 진행</b>
 ServiceTest라면 각 Service들을 이용해 테스트 해야하는데 제 머리로는
 의존성을 주입하는 @Autowired, @Inject만 생각이 나서 제대로 구현을 못했습니당.. ㅠㅜ
 
 <b>2. 하나의 ServiceTest에서 너무 다양한 책임을 사용 </b>
```java
class ResrvationServiceTest{
   private final ReservationService reservationService;
   private final ReservationTimeService reservationTimeService ;
   private final ThemeService themeService;
   ...
}
```
현재 로직 특성상 예약은 사전에 예약시간과 테마를 저장해야 하는데... 
테스트 진행시에도 위와 같이 할 수 밖엔 없을까욤? ㅠㅜ

<b>3. given의 활용 </b>
```java
class ResrvationControllerTest{
       void given_예약_시간_10_20(){
            given_예약_삭제();
            SaveReservationTimeRequest given_예약시간_10_20 = new SaveReservationTimeRequest("10:20");

            RestAssured.given().log().all()
                    .body(given_예약시간_10_20)
                    .contentType(ContentType.JSON)
                    .when().post("/times");
        }

        void given_예약_삭제(){
            RestAssured.given().log().all()
                    .when().delete("/times/1");

        }
   ...
}
```
위 처럼 given_을 구성하고 아래는 위의 given을 활용하는 방법으로 진행했습니다.
혹시 위의 방법보다 더 좋은 방법이 있을까요?

<b>4. 조회시 optional을 활용한 유효성 검증 </b>
NPE(NullPointException) 발생 방지를 위해 응답값이 존재하는 조회 메서드에 optional을
사용해보지 못해 아쉽네요.. ㅎ

OikimiO committed:
refactor: ViewController 추가 및 사용되지 않는 클래스 제거

- AdminController 제거
- 기존 Controller > RestController로 변경

OikimiO committed:
refactor: 쿼리 메서드 명명 규칙 적용

- DTO read~: find~
- <List> read~: findAll
- DTO save ~: save
- void delet ~: delete

OikimiO committed:
refactor: ReservationControllerTest & ReservationTimeControllerTest 추가

- AdminControllerTest 삭제후 각 ControllerTest추가
- Context~ > Context_use~ 변경
- method save~, return ~ > it_save~, it_return~ 변경
- given 전역 변수 > given 지역 변수로 변경

OikimiO committed:
test: Theme에서 사용할 기능 정리

OikimiO committed:
feat: Theme 사용을 위한 페이지,쿼리,스키마 추가

OikimiO committed:
feat: ThemeController&Service&Dto 구현

OikimiO committed:
feat: Theme Entity&Dto 내용 추가

OikimiO committed:
feat: ThemeService 기능 구현 및 Repositort 추가

OikimiO committed:
feat: Theme Query 추가 및 기능 사용을 위한 기능 변경

- Dto: Theme 추가&조회를 위한 기능 추가
- Reservation: Theme 조회를 위한 변수 추가
- ReservationQuery&ReservationRepository: Theme관련 조회,저장 기능 추가
- ThemeQuery&ThemeRepository: Theme 메서드 구현 및 쿼리 제작

OikimiO committed:
test: 유효성 및 검증을 위한 메서드 추가

-

OikimiO committed:
feat: spring-boot-starter-test 추가

OikimiO committed:
feat: ControllerAdvice 추가 및 예약 유효성 기능 테스트

- ErrorResponse DTO로 에러 관리

OikimiO committed:
feat:  예약시간 유효성 기능 추가 및 테스트

OikimiO committed:
feat: 테마 유효성 기능 추가 및 테스트 외 테스트 방식 변경

- 테스트시 저장값은 반드시 given_형태의 메서드 사용
- 매 테스트마다 삭제/저장 하는 번거러움을 제거하는 방안 생각해볼 것

OikimiO committed:
feat: 커스텀 Exception&test 추가

OikimiO committed:
refactor: 날짜 타입 변경 외 수정 작업

xrabcde commented on src/main/java/roomescape/admin/controller/ThemeController.java:
```diff
@@ -0,0 +1,42 @@
+import java.util.List;
+
+
+@RestController
+public class ThemeController {
```
> 해당 조언에 따라 ViewController와 비지니스Controller로 구분해보았습니다

이전보다 확실히 controller 들의 역할이 명확해진 것 같아요 ! 
[RequestMapping](https://www.baeldung.com/spring-requestmapping) 도 적용해볼 수 있을 것 같습니다

OikimiO replied:
<img width="1268" alt="image" src="https://github.com/next-step/spring-roomescape-reservation/assets/53805469/a0bae67f-1856-4eab-8b84-c82dfe0dcb78">

요청 주신 자료와 공식 문서를 함께 확인했습니당 ㅎ

### RequestMapping?
웹 요청을 처리할 메서드나 클래스에 매핑하기 위해 사용되는 스프링 애너테이션입니다.

### AS_IS > TO_BE
메서드별 Mapping 적용 > 클래스별 Mapping 적용

xrabcde commented on src/main/java/roomescape/admin/dto/SaveReservationRequest.java:
```diff
@@ -1,6 +1,14 @@
+public record SaveReservationRequest(@NotEmpty(message = "이름은 비어 있을 수 없습니다.")
+                                     @Pattern(regexp = "^[a-zA-Z가-힣]*$", message = "이름은 문자만 포함할 수 있습니다.")
+                                     String name,
+                                     @NotEmpty(message = "날짜는 비어 있을 수 없습니다.")
+                                     @Pattern(regexp = "^[0-9-]+$", message = "날짜에는 숫자와 '-'만 포함될 수 있습니다.")
```
Spring Validator 를 사용해주셨군요 👏
NotNull, NotEmpty, NotBlank 는 각각 어떻게 다를까요?

OikimiO replied:
### NotNull? 
필드가 null이 아닌지 확인하며, null이 아닌 빈 값은 허용합니다.

### NotEmpty?
필드가 null이 아니며 길이나 크기가 0보다 커야 함을 확인합니다.

### NotBlank? 
필드가 null이 아니고 공백을 제거한 길이가 0보다 커야 함을 확인합니다.

ArgumentValidationTest 클래스를 제작하여 해당 내용을 테스트를 진행하겠습니다ㅎ

xrabcde commented on src/main/java/roomescape/admin/repository/ReservationQuery.java:
```diff
@@ -1,19 +1,35 @@
 public class ReservationQuery {
-    public static final String READ_ALL = "SELECT \n" +
+    public static final String FIND_ALL = "SELECT \n" +
             "    r.id as reservation_id, \n" +
             "    r.name as reservation_name, \n" +
             "    r.date as reservation_date, \n" +
             "    t.id as time_id, \n" +
-            "    t.start_at as time_start_at \n" +
+            "    t.start_at as time_start_at, \n" +
+            "    te.id as theme_id, \n" +
+            "    te.name as theme_name, \n" +
+            "    te.description as theme_description, \n" +
```
멀티라인 String에 [text block](https://www.baeldung.com/java-text-blocks) 을 사용할 수 있을 것 같아요

OikimiO replied:
### Text Block
멀티 라인 문자열을 단순화하고 가독성을 높이는 기능을 말합니다.

### AS_IS > TO_BE
1. ~Query 멀티라인 문자열 > Text Block을 활용한 정리

xrabcde commented on src/main/java/roomescape/admin/service/ReservationService.java:
```diff
@@ -4,33 +4,74 @@
+
+        if(isUnderDate(saveReservationRequest)){
+            throw new CustomException("예약 시간은 현재 시간보다 이전일 수 없습니다.");
+        }
```
꼼꼼한 예외처리 좋습니다 👍👍

xrabcde commented on src/main/java/roomescape/admin/service/ReservationService.java:
```diff
@@ -4,33 +4,74 @@
+            return true;
+        }
+
+        return false;
```
바로 return 해줄 수 있을 것 같아용

```suggestion
        return reserve.isBefore(LocalDateTime.now());
```

xrabcde commented on src/main/java/roomescape/admin/service/ReservationService.java:
```diff
@@ -4,33 +4,74 @@
+            return true;
+        }
+
+        return false;
```
여기두용

xrabcde commented on src/main/java/roomescape/admin/service/ThemeService.java:
```diff
@@ -0,0 +1,35 @@
+        return ReadThemeResponse.from(theme);
+    }
+
+    public void deleteTheme(Long id) {
+        this.themeRepository.delete(id);
+    }
```
> 예약(Reservation)은 예약시간(ReservationTime)과 테마(Theme)에 대한 외래키를 가집니다.
이로 인해 예약목록 중 예약시간,테마를 사용하고 있다면 외래키에 해당하는 값은 삭제할 수 없습니다.
해당 부분은 ... 당시엔 생각이 안나... 예외 처리를 진행하지 못했습니다..
생각을 정리하며 생각난건데.. 예외 처리를 Exception이 발생하는 영역에
try with resource 형태로 구현하면 어떨까 싶습니다.

테마나 시간을 삭제하기 전에 삭제 대상과 관련된 예약이 있는지 체크해보는 건 어떤가요?

OikimiO replied:
계속 어떻게 해야되나 고민했는데.. ㅎㅎ 감사합니다 ㅎ

xrabcde commented on src/main/java/roomescape/common/exception/CustomException.java:
```diff
@@ -0,0 +1,8 @@
+package roomescape.common.exception;
+
+public class CustomException extends RuntimeException{
+
+    public CustomException(String message){
+        super(message);
```
커스텀 예외를 활용해주셨군요! 
커스텀 예외는 왜 사용할까요? 이 예외의 이름은 어떤 의미를 담고 있나요?

OikimiO replied:
### 현재의 CustomException은 어떤 의미인가요?
현재의 CustomException은 모든 Exception에서 예외를 처리하겠다는 의미로 작성되었습니다.

### 커스텀 예외 왜 사용할까요? 
1. 명확한 예외 처리
2. 비지니스 로직 반영
3. 가독성 향상
4. 예외 계층 구조

### AS_IS > TO_BE
1. 모든 Exception에서 예외 처리 > 비지니스의 정책 예외를 처리하는 Exception 예외로 변경

### 궁금증
1. 일반적으로 실무에선 너무 다양한 Exception들이 존재합니다. 
만약 커스텀 예외의 의도가 <b>명확한 예외 처리</b>라면 실무에선 어떤 범주로 커스텀 Exception을 나눌 수 있을까요?

2. Exception의 범주가 다양해지면 관리가 어려워 하나로 관리하는게 좋다는 의견을 받은 적이 있습니다.  
만약 필요에 따라 다양한 Exception이 만들어 졌다면 그 흐름을 하나로 관리하는 방법이 있을까요?

xrabcde commented on src/test/java/roomescape/admin/controller/ReservationControllerTest.java:
```diff
@@ -0,0 +1,216 @@
+import java.util.List;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.DEFINED_PORT)
+public class ReservationControllerTest {
+
```
> ControllerTest의 역할
각 Controller 호출을 통한 프로세스 흐름 파악
Controller에서 저장 호출시 데이터 유효성 검증 파악

> ServiceTest의 역할
각 도메인별 정책 파악

각 테스트 별 역할까지 고민해보셨군요 💯💯
명확한 테스트를 위해 많이 고민하신 게 코드에서 느껴집니다 :)

OikimiO replied:
### ServiceTest의 역할 추가
AS_IS > TO_BE
1. 각 도메인별 정책 파악 > 각 도메인별 정책 파악 + DB CRUD 예외 관리

xrabcde commented on src/test/java/roomescape/admin/controller/ReservationControllerTest.java:
```diff
@@ -0,0 +1,216 @@
+                assertThat(response.statusCode()).isEqualTo(HttpStatus.BAD_REQUEST.value());
+            }
+
+            @Test
+            @DisplayName("date에 null, (공백) 값이 입력되면, 에러를 리턴합니다.")
+            void it_return_error_for_null_or_empty_date(){
```
예외 케이스 테스트까지 👍💯

xrabcde commented on src/test/java/roomescape/admin/controller/ReservationTimeControllerTest.java:
```diff
@@ -0,0 +1,119 @@
+            assertThat(response.jsonPath().getLong("id")).isPositive();
+        }
+
+        @Test
+        @DisplayName("목록 List를 리턴합니다.")
+        void it_return_reservation_times()  {
```
테스트를 너무 꼼꼼히 잘 짜주셨는데 시간이 촉박해서 그런지 실패하는 케이스가 몇 개 있는 것 같아요 
확인 부탁드립니다 !! 

<img width="713" alt="image" src="https://github.com/next-step/spring-roomescape-reservation/assets/56033755/c0a4df84-550d-4f3a-baef-e9aaf25d00b9">

OikimiO replied:
헷.. 참고하겠습니당 ㅎ

xrabcde replied:
여전히 전체 테스트 실행시, 목록 조회 테스트가 깨지고 있어요 🥲
개별테스트 시에는 성공하지만, 전체 테스트에서만 실패하는 이유가 뭘까요? 어떻게 개선할 수 있을까요? 
고민해보시고 다음 미션부터 적용해보시면 좋을 것 같습니다 😉

<img width="956" alt="image" src="https://github.com/next-step/spring-roomescape-reservation/assets/56033755/22d6d502-bb7b-4232-803c-df57a91d727e">

xrabcde commented on src/test/java/roomescape/admin/service/ReservationServiceTest.java:
```diff
@@ -0,0 +1,136 @@
+import roomescape.admin.dto.*;
+import roomescape.admin.entity.ReservationTime;
+import roomescape.admin.entity.Theme;
+import roomescape.admin.repository.ReservationRepository;
```
사용하지 않는 import 문은 정리해주세요 ~

xrabcde commented on src/test/java/roomescape/admin/service/ReservationServiceTest.java:
```diff
@@ -0,0 +1,136 @@
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Mockito.mock;
+
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.DEFINED_PORT)
+public class ReservationServiceTest {
```
> 현재 로직 특성상 예약은 사전에 예약시간과 테마를 저장해야 하는데...
테스트 진행시에도 위와 같이 할 수 밖엔 없을까욤? ㅠㅜ

Reservation이 테마와 시간에 대한 연관관계를 갖고 있어서 테스트 데이터가 고민이신 것 같아요 
이렇게 연관관계를 갖는 도메인에서는 테스트 데이터 관리가 늘 숙제이긴 합니다 🥲
도메인이 복잡해지면 test fixture 를 도입하는 방법도 고려해보실 수 있는데요 
이건 필수 요구사항은 아니니 은호님 선택에 맡기도록 하겠습니다 ☺️

xrabcde commented on src/test/java/roomescape/admin/service/ThemeServiceTest.java:
```diff
@@ -0,0 +1,104 @@
+            @Test
+            @DisplayName("삭제할 수 없고, 에러를 리턴합니다.")
+            void it_return_error_and_not_delete_theme(){
+                given_예약_2024_12_12();
+
+                var response = RestAssured
+                        .given().log().all()
```
> 위 처럼 given_을 구성하고 아래는 위의 given을 활용하는 방법으로 진행했습니다.
혹시 위의 방법보다 더 좋은 방법이 있을까요?

네 given 을 사용하는 현재 테스트 코드도 읽기 쉽게 잘 작성해주셔서 저는 괜찮은 것 같습니다 :)

다만 현재는 ServiceTest 에서도 api 호출을 통해 로직을 확인해주고 계신데 
Service 의 역할을 테스트 하는 것이 목적이라면 api 호출이 아닌 
서비스 메서드를 호출하는 방향으로 변경해보시는 게 어떨까요?

OikimiO replied:
### Service Test 적용
```java 
public class ServiceTest {
     private final ReservationService reservationService;

    public ATest(ApplicationContext context) {
        this.reservationService = context.getBean(ReservationService.class);
    }

   ... 
}
```

xrabcde commented on src/test/java/roomescape/admin/service/ThemeServiceTest.java:
```diff
@@ -0,0 +1,104 @@
+            void given_테마(){
+                SaveThemeRequest 테마 = new SaveThemeRequest("테마","테마설명","썸네일 URL");
+
+                var response = RestAssured.given().log().all()
```
요건 사용하지 않고 있는 걸까용?

xrabcde reviewed (changes requested):
안녕하세요 은호 님 ~
2단계도 요구사항에 맞게 잘 구현해주셨네요 ㅎㅎ
몇 가지 코멘트 남겨드렸는데 확인해보시고 다시 리뷰요청 부탁드립니다 🙇🏻‍♀️

OikimiO committed:
refactor: RequestMapping을 활용해 class 매핑으로 변경

OikimiO committed:
refactor: 각 쿼리문에 text block 적용 및 예약 정책을 위한 추가

- 테마 혹은 예약시간 예약에서 사용된다면 삭제할 수 없는 쿼리문 추가

OikimiO committed:
refactor: Exception 추가/수정 및 메서드 수정

- CustomException > PolicyException으로 변경
- DBException 추가
- ~Service: 각 서비스 조회에 Optional 추가
- ~Service: 테마, 예약시간이 예약에 사용될 경우 삭제 못하는 정책 적용
- ControllerAdvice에  변경된 Exception 클래스가 사용되도록 수정

OikimiO committed:
test: service ApplicationContext를 통한 Bean 가져온 후 테스트 진행

OikimiO committed:
test: 인자값 유효성 테스트 추가

OikimiO committed:
test: MissionStepTest 테스트 일부 수정

xrabcde commented on src/main/java/roomescape/admin/controller/ReservationController.java:
```diff
@@ -10,28 +10,29 @@
 import java.util.List;
 
 @RestController
+@RequestMapping(value = "/reservations")
```
👍👍

xrabcde commented on src/main/java/roomescape/admin/service/ThemeService.java:
```diff
@@ -17,19 +22,43 @@ public ThemeService(ThemeRepository themeRepository){
     }
 
     public void deleteTheme(Long id) {
+        if(isContainReservation(id)){
+            throw new PolicyException(THEME_IN_USE_RESERVATION_ERROR);
+        }
+
```
삭제 전 예외처리 좋습니다 ㅎㅎ 잘 동작하네요 !

xrabcde reviewed (approved):
안녕하세요 은호님 ~ 피드백을 잘 반영해주셨네요 ㅎㅎ
첫 주차 미션 진행하시느라 고생많으셨습니다
몇 가지 코멘트 남겨드렸는데 시간 되실 때 확인해주세요 😉
이어서 다음 미션들도 화이팅입니다 ~!
